### PR TITLE
🏁  CML AMP Integration Testing

### PR DESCRIPTION
**Pull request by andrewrreed**
# Purpose

The purpose of this PR is to introduce, integrate, and validate all necessary artifacts required make this repo compatible with CML AMP declarative launch scripts.

# Changes made

- Update `README.md` to specify that if running manually on CML you need at least 4 GB RAM
- Created a new directory `/cml` at root of project to hold CML launch artifacts (following [Victor's lead](https://github.com/fastforwardlabs/imageanalysis_cml))
- Moved `install_dependencies.py` to the `/cml` folder. This file is necessary to create a job that kicks off dependency installation step via the `.project-metadata.yaml`
- Created `/cml/fit_models_parallel.py` that utilizes the [CDSW Workers API](https://docs.cloudera.com/machine-learning/cloud/distributed-computing/topics/ml-workers-api.html) to launch each model training script in parallel
- Updated the `.project-metadata.yaml` file to successfully execute the AMP build process on CML

# What to test

The following tests have been validated before submitting this PR. However, before calling this complete, these should probably be tested again by someone other than myself (@cjwallace 🙏 ):

- [x]  The AMP successfully executes each step in when launched on CML via `ML Prototype` as the Initial Setup option
- [x]  The AMP successfully builds all data artifacts
- [x]  The AMP successfully launches both applications
- [x]  The Title, Short Description, Long Description, Tags, CML Features fields are all complete and correct in the [AMP Tracking doc](https://docs.google.com/spreadsheets/d/1Z3BSQSwAoF4TIjguFVDZVXk96h9n30jscFfI0YFTjO0/edit#gid=0)

# How to test

This is a bit tricky because the current ML Prototype functionality on CML does not allow you to specify which branch to load and run, nor does it respect the `default` branch of the git repo. For that reason, it might be best to make an exception for now and just merge these changes, then test (and revert if needed). We can also merge and walk through the tests together as well (that way we can fix any issues on the spot, if there are any...)

1. Log into a [CML cluster that contains the ML Prototype functionality](https://ml-3c960a83-7fb.eng-ml-q.svbr-nqvp.int.cldr.work/)
2. Create a new project with Initial Setup == `ML Prototype` and provide the link to this git repo (after changes have been merged 😐 )
3. Create Project --> Configure Project

# What a successful run looks like

image
image
image

### CODE REVIEW

**Comment by andrewrreed** [comment]
**Note** - While testing it was discovered that the project dependencies fail to install unless the session on which the `pip3 install` is run has a minimum of: *2 vCPU / 4 GiB Memory*

If a smaller session instance is used to install dependencies, the following error occurs:

image


**ACTION** - Adding a note into the projects `README.md` in this PR to inform the user.

**Comment by andrewrreed** [comment]
@cjwallace Opening this PR up for your review whenever you have the chance. Feel free to provide feedback here or make updates/edits/contributions as you see fit. Happy to walk through this with you as well because the testing piece isn't the most straightforward (as mentioned in the PR body). Thanks!

**Comment by victordibia** [comment]
Fine work @andrewrreed! 

> This is a bit tricky because the current ML Prototype functionality on CML does not allow you to specify which branch to load and run, nor does it respect the default branch of the git repo

One thing I have done in the past is to make a fork of the repo (instead of a branch) and use that to test my changes.

File: .project-metadata.yaml [hunk 1]
```diff
@@ -13,18 +13,55 @@ environment_variables:
     prompt_user: false
 
 tasks:
+  - type: create_job
+    name: Install Dependencies
+    entity_label: install_dependencies
+    script: cml/install_dependencies.py
+    arguments: None
+    cpu: 1
```
Comment (victordibia) [inline .project-metadata.yaml]: You can change CPU to 2 here if this is needed.

Comment (victordibia) [inline .project-metadata.yaml]:
```suggestion
    cpu: 2
```

**Comment by andrewrreed** [comment]
> Fine work @andrewrreed!
> 
> > This is a bit tricky because the current ML Prototype functionality on CML does not allow you to specify which branch to load and run, nor does it respect the default branch of the git repo
> 
> One thing I have done in the past is to make a fork of the repo (instead of a branch) and use that to test my changes.

Good call - this is definitely the better way to do it. My only problem with that approach (in this instance) is that I already have [a fork of the repo](https://github.com/andrewrreed/structural-time-series-kfp) that I have merged all of the Kubeflow Pipelines work into `master` on. In hindsight, I guess the proper workaround would have been to make a temporary copy of the `master` branch in my forked repo, reversed commits to be inline with this repo, and then built from there..

**Comment by cjwallace** [comment]
Pro-tip. If testing on nightly, need to enable a 2vCPU, 4GB container, or it'll hang indefinitely. With that resolved, it works! Great job @andrewrreed. I'll take care of updating the tracking doc and squashing the commits for merge tomorrow.

**Review by cjwallace — APPROVED** [review]

**Pull request merged**